org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/ClassfileContentHandler.java

```diff
@@ -16,36 +16,60 @@
 import org.eclipse.jdt.core.JavaModelException;
 import org.eclipse.jdt.ls.core.internal.JDTUtils;
 import org.eclipse.jdt.ls.core.internal.JavaLanguageServerPlugin;
+import org.eclipse.jdt.ls.core.internal.managers.DecompilerManager;
+import org.eclipse.jdt.ls.core.internal.preferences.PreferenceManager;
 import org.eclipse.lsp4j.TextDocumentIdentifier;
 
 public class ClassfileContentHandler {
 
 	private static final String EMPTY_CONTENT = "";
 
+	private final PreferenceManager preferenceManager;
+
+	public ClassfileContentHandler(PreferenceManager preferenceManager) {
+		this.preferenceManager = preferenceManager;
+	}
+
 	public String contents(TextDocumentIdentifier param, IProgressMonitor monitor) {
 		String source = null;
+		IClassFile cf = JDTUtils.resolveClassFile(param.getUri());
+		if (cf != null) {
+			source = contents(cf, monitor);
+		}
+		if (source == null) {
+			source = EMPTY_CONTENT;// need to return non null value
+		}
+		return source;
+	}
+
+	public String contents(IClassFile cf, IProgressMonitor monitor) {
+		String source = null;
 		try {
-			IClassFile cf = JDTUtils.resolveClassFile(param.getUri());
-			if (cf != null) {
-				IBuffer buffer = cf.getBuffer();
-				if (buffer != null) {
-					if (monitor.isCanceled()) {
-						return EMPTY_CONTENT;
-					}
-					source = buffer.getContents();
-					JavaLanguageServerPlugin.logInfo("ClassFile contents request completed");
-				}
-				if (source == null) {
-					source = JDTUtils.disassemble(cf);
+			IBuffer buffer = cf.getBuffer();
+			if (buffer != null) {
+				if (monitor.isCanceled()) {
+					return EMPTY_CONTENT;
 				}
+				source = buffer.getContents();
+				JavaLanguageServerPlugin.logInfo("ClassFile contents request completed");
+			}
+			if (source == null) {
+				source = decompile(cf, monitor);
 			}
 		} catch (JavaModelException e) {
 			JavaLanguageServerPlugin.logException("Exception getting java element ", e);
 		}
-		if (source == null) {
```

Comment:
either this null-check needs to stay, or decompile(...) needs to guarantee non-null values (eg. in case of exception)

Reply:
Not so—this method is called in two places:

1. By the `contents` method above, which has the original null check
2. By `JDTUtils.findElementsAtSelection`, which requires a null if no content was found